sound/soc/sof/ipc4-topology.c

```diff
@@ -943,6 +944,21 @@ static int sof_ipc4_widget_setup_comp_process(struct snd_sof_widget *swidget)
 
 	sof_ipc4_widget_update_kcontrol_module_id(swidget);
 
+	pipe_widget = swidget->spipe->pipe_widget;
+
+	/*
+	 * A Data Processing (DP) domain widget can be scheduled to run on a secondary core while
+	 * it's pipeline runs the primary core. In this case, the secondary core could get powered
+	 * off before the pipeline is deleted resulting in a crash when the firmware tries to free
+	 * the DP module during pipeline deletion. Prevent this by setting the pipe_widget's core
+	 * ID to match that of the widget so that the core would be kept powered on until the
+	 * pipeline has been deleted. This would work only in the case where a pipeline only
+	 * contains a single DP module that is scheduled on a secondary core.
```

Comment:
This is not great....

At the very least we would need to have something that detects that there are more than 2 non-primary cores used by a pipeline. But clearly with the AEC/NS sequence it's quite likely that this configuration will happen.

But I would really not invest any time in this, it's a stop-gap which will be defeated by 3rd parties.

Reply:
I agree, a pipeline with couple of DPs spread on all cores will be needed probably very soon.